src/memory_prefetch.c

```diff
@@ -8,15 +8,282 @@
  */
 
 #include "memory_prefetch.h"
+#include "server.h"
+
+typedef enum {
+    PREFETCH_ENTRY, /* Initial state, prefetch entries associated with the given key's hash */
+    PREFETCH_VALUE, /* prefetch the value object of the entry found in the previous step */
+    PREFETCH_DONE   /* Indicates that prefetching for this key is complete */
+} PrefetchState;
+
+typedef struct KeyPrefetchInfo {
+    PrefetchState state; /* Current state of the prefetch operation */
+    hashtableIncrementalFindState hashtab_state;
+} KeyPrefetchInfo;
+
+/* PrefetchCommandsBatch structure holds the state of the current batch of client commands being processed. */
+typedef struct PrefetchCommandsBatch {
+    size_t cur_idx;                 /* Index of the current key being processed */
+    size_t keys_done;               /* Number of keys that have been prefetched */
+    size_t key_count;               /* Number of keys in the current batch */
+    size_t client_count;            /* Number of clients in the current batch */
+    size_t max_prefetch_size;       /* Maximum number of keys to prefetch in a batch */
+    size_t executed_commands;       /* Number of commands executed in the current batch */
+    int *slots;                     /* Array of slots for each key */
+    void **keys;                    /* Array of keys to prefetch in the current batch */
+    client **clients;               /* Array of clients in the current batch */
+    hashtable **keys_tables;        /* Main table for each key */
+    hashtable **expire_tables;      /* Expire table for each key */
+    KeyPrefetchInfo *prefetch_info; /* Prefetch info for each key */
+} PrefetchCommandsBatch;
+
+static PrefetchCommandsBatch *batch = NULL;
+
+void freePrefetchCommandsBatch(void) {
+    if (batch == NULL) {
+        return;
+    }
+
+    zfree(batch->clients);
+    zfree(batch->keys);
+    zfree(batch->keys_tables);
+    zfree(batch->expire_tables);
+    zfree(batch->slots);
+    zfree(batch->prefetch_info);
+    zfree(batch);
+    batch = NULL;
+}
 
 void prefetchCommandsBatchInit(void) {
+    serverAssert(!batch);
+    size_t max_prefetch_size = server.prefetch_batch_max_size;
+
+    if (max_prefetch_size == 0) {
+        return;
+    }
+
+    batch = zcalloc(sizeof(PrefetchCommandsBatch));
+    batch->max_prefetch_size = max_prefetch_size;
+    batch->clients = zcalloc(max_prefetch_size * sizeof(client *));
+    batch->keys = zcalloc(max_prefetch_size * sizeof(void *));
+    batch->keys_tables = zcalloc(max_prefetch_size * sizeof(hashtable *));
+    batch->expire_tables = zcalloc(max_prefetch_size * sizeof(hashtable *));
```

Comment:
What are the scenarios/conditions that trigger access to the expire table?

Reply:
The expire table needs to be accessed only to insert, delete or replace an object.

For example in GET, if a key is expired, the command needs to access the expire table to delete it. So maybe it's good to do it, but OTOH, if this unlikely to happen, then we evict other stuff from the L1 cache when we add this.

I'm not sure, so let's leave it for now?

Reply:
Another possibility is that we can do it when we have the robj AKA valkey object. Here we can actually check the expire value in this prefetch code.

Reply:
Thanks, I will run some tests with expired keys to check what give the best result.

Reply:
I think another option would be to interleave the expire table scan with the hashtable scan in case the command is a mutational command (as @zuiderkwast suggested). although we do not know the key has expire or not, waiting for the key and only then start the expiry scan seems like something we can optimize via parallelization.
In most cases I have seen either the whole dataset has TTL or only a minor fraction. maybe we can use that assumption and optimize since most workloads do have expiry on objects (for volatile eviction). Maybe we can test if tit improves later on @uriyage

Reply:
Current code performance tests results:

| Scenario | Operation | With Prefetch (K ops/sec) | Without Prefetch (K ops/sec) | Change |
|----------|-----------|---------------------------|-----------------------------| -------|
| All Keys Expired | SET | 747 | 697 | +7% |
| All Keys Expired | GET | 985 | 930 | +5.9% |
| 50% Expired | GET | 983 | 985 | ~0% |
| 1% Expired | GET | 1049 | 1102 | -4.8% |

The results show that prefetching behavior has different impacts depending on the percentage of expired keys. While it improves performance when all keys have expiry (+7% for SET, +5.9% for GET), it actually hurts performance by 4.8% when only 1% of keys are expired.

> For example in GET, if a key is expired, the command needs to access the expire table to delete it. So maybe it's good to do it

The optimization to access the expire table only when an object has actually expired is still in `TODO`

After modifying the code to check the expiration time before accessing the expire table
```c
-         /* TODO: Use a better design for expireIfNeeded, now that the expire is
-          * stored in the value, for example passing val to expireIfNeeded. */
-        if (objectGetExpire(val) != -1 && expireIfNeeded(db, key, expire_flags) != KEY_VALID) {
-
+        int expired = 0;
+        mstime_t when = objectGetExpire(val)
+
+        if (when > 0) {
+            mstime_t now = commandTimeSnapshot();
+            
+            if (now > when) {
+                expired = 1;
+            }
+        }
+
+        if (expired && expireIfNeeded(db, key, expire_flags) != KEY_VALID) {
             /* The key is no longer valid. */
             val = NULL;
         }
```
here are the new results:

| Operation | With Prefetch (K ops/sec) | Without Prefetch (K ops/sec) | Change |
|-----------|---------------------------|-----------------------------| -------|
| SET | 765 | 800 | -4.4% |
| GET | 1020 | 1115 | -8.5% |

Given these results, I'm convinced now we should remove the expire table prefetch. With the optimization to check the expiration time before accessing the table, we'll only need to access it when an object has actually expired. This change shows significant performance improvements, especially for GET.

Note that SET operations also show better performance as currently we check for `expireIfNeeded` despite the key being about to be overwritten.